packages/lesswrong/server/analytics/analyticsCron.ts

```diff
@@ -0,0 +1,115 @@
+      inclusiveEndDate === null ||
+      currentDate < earliestWindowStart ||
+      currentDate >= inclusiveEndDate
+    ) {
+      const startOfDay = moment(currentDate).utc().startOf("day").toDate();
+      const endOfDay = moment(currentDate).utc().endOf("day").toDate();
+      ranges.push({ startDate: startOfDay, endDate: endOfDay });
+    }
+    // Move to the next day
+    currentDate = moment(currentDate).utc().add(1, 'day').toDate();
+  }
+
+  // Reverse ranges so the most recent days are done first
+  ranges.reverse()
+
+  for (const range of ranges) {
+    // Allow individual days to fail
+    try {
+      console.log(`Updating data for range: ${range.startDate.toISOString()} to ${range.endDate.toISOString()}`)
+
+      const data = await repo.calculateDataForDateRange(range);
+
+      console.log(`Calculated ${data.length} rows, upserting data`)
+
+      await repo.upsertData({ data })
+      console.log(`Finished updating data for range: ${range.startDate.toISOString()} to ${range.endDate.toISOString()}`)
+    } catch (e) {
+      console.error(e)
+      continue
+    }
+  }
+}
+
+async function updatePostViews({earliestStartDate, latestEndDate}: {earliestStartDate: Date, latestEndDate: Date}) {
+  console.log("Updating PostViews collection")
+  await updateDailyAnalyticsCollection({repo: new PostViewsRepo(), earliestStartDate, latestEndDate})
+  console.log("Finished PostViews collection")
+}
+
+async function updatePostViewTimes({earliestStartDate, latestEndDate}: {earliestStartDate: Date, latestEndDate: Date}) {
+  console.log("Updating PostViewTimes collection")
+  await updateDailyAnalyticsCollection({repo: new PostViewTimesRepo(), earliestStartDate, latestEndDate})
+  console.log("Finished PostViewTimes collection")
+}
+
+async function updateAnalyticsCollections(props: {startDate?: string}) {
+  const { startDate } = props ?? {}
+
+  // If no explicit start date is given, only go back 1 week to avoid this being slow
+  const earliestStartDate = startDate ? moment(startDate).utc().startOf('day').toDate() : moment().utc().subtract(1, 'week').startOf('day').toDate();
+  // endDate is the end of the current day
+  const endDate = moment().utc().endOf('day').toDate();
+
+  console.log(`Starting updateAnalyticsCollections. startDate (given): ${startDate}, earliestStartDate: ${earliestStartDate}, endDate: ${endDate}`)
+
+  try {
+    await updatePostViews({earliestStartDate, latestEndDate: endDate})
+  } catch (e) {
+    console.error("Failed to update PostViews collection")
+    console.error(e)
+  }
+
+  try {
+    await updatePostViewTimes({earliestStartDate, latestEndDate: endDate})
+  } catch (e) {
+    console.error("Failed to update PostViewTimes collection")
+    console.error(e)
+  }
+
+  console.log("Finished updateAnalyticsCollections")
+}
+
+if (isEAForum) {
+  addCronJob({
+    name: "updateAnalyticsCollections",
+    interval: "every 20 minutes",
+    job: async () => updateAnalyticsCollections({}),
```

Comment:
Do you mean why it's defined at the default in the function rather than here? You could do it either way 🤷‍♂️, one reason to prefer this is like "it's simpler if there's no logic in the cron job, so you can keep track of all the logic", but idk